### ISSUE TITLE
kops fatal error: found duplicate tasks with name "IAMRole"

### ISSUE THREAD

**Issue by kesor** [kubernetes/kops#9852]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

`v1.17.0` , `v1.17.1` , `v1.18.0`

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

`v1.17.11`

**3. What cloud provider are you using?**

Amazon AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

```shell
kops create cluster
kops create ig second
# add `iam:\nprofile: arn:aws:iam::123456789012:instance-profile/nodes.mycluster.domain.com`
# the IAM profile used for second IG is the same as `nodes`
kops update cluster
```

**5. What happened after the commands executed?**

A fatal error appears:

```
F0901 14:58:45.758233   63950 task.go:59] found duplicate tasks with name "IAMRole/nodes.mycluster.domain.com": *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"} and *awstasks.IAMRole {"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"}
```

**6. What did you expect to happen?**

To be able having a mix of explicit `iam: profile:` instance groups, and some that are implicit. At the moment `kops` allows to share IAM InstanceProfile across multiple instance groups, but either ALL have to have `iam: profile:` be explicit, or ALL have to have `iam: profile:` be implicit.

**7. Please provide your cluster manifest. Execute `kops get --name my.example.com -o yaml` to display your cluster manifest. You may want to remove your cluster name and other sensitive information.**

```yaml
apiVersion: kops.k8s.io/v1alpha2
kind: Cluster
metadata:
  creationTimestamp: "2020-09-01T11:33:17Z"
  generation: 1
  name: mycluster.domain.com
spec:
  api:
    loadBalancer:
      type: Public
  authorization:
    rbac: {}
  channel: stable
  cloudLabels:
    k8sClusterName: mycluster
  cloudProvider: aws
  configBase: s3://mybucket.domain.com/mycluster.domain.com
  dnsZone: domain.com
  etcdClusters:
  - cpuRequest: 200m
    etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    memoryRequest: 100Mi
    name: main
  - cpuRequest: 100m
    etcdMembers:
    - instanceGroup: master-us-east-1a
      name: a
    memoryRequest: 100Mi
    name: events
  iam:
    allowContainerRegistry: true
    legacy: false
  kubelet:
    anonymousAuth: false
  kubernetesApiAccess:
  - 0.0.0.0/0
  kubernetesVersion: 1.17.11
  masterInternalName: api.internal.mycluster.domain.com
  masterPublicName: api.mycluster.domain.com
  networkCIDR: 172.31.0.0/16
  networking:
    canal: {}
  sshAccess:
  - 0.0.0.0/0
  subnets:
  - cidr: 172.31.2.0/27
    name: us-east-1a
    type: Private
    zone: us-east-1a
  - cidr: 172.31.1.0/27
    name: utility-us-east-1a
    type: Utility
    zone: us-east-1a
  topology:
    dns:
      type: Private
    masters: private
    nodes: private
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

https://gist.github.com/kesor/953d33ce28f1563e55072e4eb91bafe2

**Comment by kesor** [comment]
I suspect this is related to code introduced in #4171 and #9693 seems related somehow, but could not understand the problem from it.

**Comment by johngmyers** [comment]
I believe https://github.com/kubernetes/kops/blob/8a81d94c7bd812d3e196c4bb14ac241e4207f752/pkg/model/iam.go#L118 needs to instead use `c.EnsureTask()`. And all the other `AddTask()` in that function need to be changed as well.

**Comment by kesor** [comment]
Based on @johngmyers suggestion, I compiled a version where all the `AddTask` were replaced with `EnsureTask` in `model/iam.go`. The result output looks like this --

```
W0903 14:24:40.850138   13220 task.go:78] EnsureTask found task mismatch for "IAMInstanceProfile/nodes.mycluster.domain.com"
W0903 14:24:40.850145   13220 task.go:79]       Existing: *awstasks.IAMInstanceProfile {"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","ID":null,"Shared":true}
W0903 14:24:40.850166   13220 task.go:80]       New: *awstasks.IAMInstanceProfile {"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","ID":null,"Shared":false}
W0903 14:24:40.850174   13220 task.go:78] EnsureTask found task mismatch for "IAMInstanceProfileRole/nodes.mycluster.domain.com"
W0903 14:24:40.850177   13220 task.go:79]       Existing: *awstasks.IAMInstanceProfileRole {"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","InstanceProfile":{"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","ID":null,"Shared":true},"Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"}}
W0903 14:24:40.850193   13220 task.go:80]       New: *awstasks.IAMInstanceProfileRole {"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","InstanceProfile":{"Name":"nodes.mycluster.domain.com","Lifecycle":"Sync","ID":null,"Shared":false},"Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"}}
W0903 14:24:40.850263   13220 task.go:78] EnsureTask found task mismatch for "IAMRolePolicy/node-policyoverride"
W0903 14:24:40.850267   13220 task.go:79]       Existing: *awstasks.IAMRolePolicy {"ID":null,"Lifecycle":"Sync","Name":"node-policyoverride","Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"},"PolicyDocument":null,"ExternalPolicies":null,"Managed":true}
W0903 14:24:40.850273   13220 task.go:80]       New: *awstasks.IAMRolePolicy {"ID":null,"Lifecycle":"Sync","Name":"node-policyoverride","Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"},"PolicyDocument":null,"ExternalPolicies":null,"Managed":true}
W0903 14:24:40.850306   13220 task.go:78] EnsureTask found task mismatch for "IAMRolePolicy/additional.nodes.mycluster.domain.com"
W0903 14:24:40.850311   13220 task.go:79]       Existing: *awstasks.IAMRolePolicy {"ID":null,"Lifecycle":"Sync","Name":"additional.nodes.mycluster.domain.com","Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"},"PolicyDocument":{"Name":"","Resource":""},"ExternalPolicies":null,"Managed":false}
W0903 14:24:40.850321   13220 task.go:80]       New: *awstasks.IAMRolePolicy {"ID":null,"Lifecycle":"Sync","Name":"additional.nodes.mycluster.domain.com","Role":{"ID":null,"Lifecycle":"Sync","Name":"nodes.mycluster.domain.com","RolePolicyDocument":{"Name":"","Resource":{}},"ExportWithID":"nodes"},"PolicyDocument":{"Name":"","Resource":""},"ExternalPolicies":null,"Managed":false}
```

**Comment by kesor** [comment]
Apparently the same IAM Instance Profile appears in both "shared" and in "managed" profile lists, and then it is being created once with the value of `Shared: true` and once with the value of `Shared: false` which causes the conflict.

https://github.com/kubernetes/kops/blob/8a81d94c7bd812d3e196c4bb14ac241e4207f752/pkg/model/iam.go#L73-L92

**Comment by kesor** [comment]
I believe this was fixed for us by #9867
